### ISSUE TITLE
Bridger internet issues when roaming between multiple APs with bridger installed

### ISSUE THREAD

**Issue by gssjshark** [nbd168/bridger#6]
I have a MT6000 as my main router and 3 MT6000 as dumb APs. It seems something broke with bridger in the 2024.04.15 update. I am having issues with the internet not working when I’m bouncing between multiple APs around the house, even though the AP shows I’m connected with a great signal.  When I remove bridger from the dumb APs and turned WED off on all units, everything works perfectly. Fast roaming works great, and I don’t loose internet connectivity when I’m bouncing around the house. Been up 24 hours and no issues, but with Bridger installed on the dumb APs and WED enabled or disabled on all the units… it lasts 5-10 mins before I lose internet walking around the house. I also turned off fast roaming and that made no difference. I am using r26047 on the main router, and r26060 on my dumb APs and was using Bridger 2024.04.22~40b1c5b6. Kernel 6.1.86 with the default firmware on all MT600.

**Comment by nxhack** [comment]
I am exactly the same with you.

I had a problem when moving between two dumb APs. Then I restarted the bridger of the two APs and the problem was fixed.

However, the same action was required if I moved again.

**Comment by gssjshark** [comment]
I have the same issue, and restating the AP solves the issue. As soon as I start moving from dumb ap to dumb ap, this issue occurs again. I hope we can help and get this issue resolved.

**Comment by APCCV** [comment]
Same (or similar) issue. Same setup as that above.
It takes 2-3 days for the issues to appear, not minutes.
Devices connected to the AP having issues lose all connectivity. If they roam to a different AP they regain network access, and if they roam back to an affected AP they lose connectivity again.
The affected AP remains online and can be reached via ethernet. Restarting WiFi (wifi down && wifi up) restores normal operation without having to reboot the router which suggest the issue may be with WiFi driver (and WED) or routing.
Routing tables always looked OK.
Nothing useful appeared in logs.
Disabling WED but leaving bridger on seemed to also trigger the issue although usually both were being used at the same time.

**Comment by gssjshark** [comment]
@nbd168 would love to get your thoughts on this and how we can help! Thank you.

**Comment by CharlesMengCA** [comment]
Similar issue on ASUS TUF-AX4200. built with bridger included, fast roaming stop working. Removed bridger, the fast roaming works again.

Looks like after a client disconnected from an AP, the bridger or something on the AP still sends the MAC alive token to the DHCP server. When the client try to get IP from the DHCP server from another AP, since the MAC still assigned IP from the old AP. the new request failed.

**Comment by CharlesMengCA** [comment]
Strange enough, my other router LINKSYS E8450 has bridger installed and the fast roaming is working fine.

**Comment by Halry** [comment]
I have this issues too. Problem will surface if you switch between APs a few times. Then the internet connection will break(at least very slow).

**Comment by gssjshark** [comment]
> I have this issue too. Problem will surface if you switch between APs a few times. Then the internet connection will break(at least very slow).

Exactly what happens to me. Thank you for sharing!

**Comment by gssjshark** [comment]
@nbd168 what would be helpful for you to look into this (if possible). Thank you!

**Comment by gssjshark** [comment]
> Strange enough, my other router LINKSYS E8450 has bridger installed and the fast roaming is working fine.

What version of bridger are you using?

**Comment by CharlesMengCA** [comment]
My bad. Bridger only works when it is not in the middle of the topology chain.

Working:
Main router (dhcp server)->AX4200->E8450 (bridger)

Not working:
Main router (dhcp server->E8450 (bridger))->AX4200

In the not working scenario, even my iPhone disconnected from AX4200, the bridger on E8450 still sent keep-alive to my main router and made it keep the lease in the dhcp table. had to reboot the E8450 to make the lease go away.

**Comment by gssjshark** [comment]
@nbd168 i was hoping to get your help with this bug! Please let us know if anything else is needed from us to help. Thank you

**Comment by gssjshark** [comment]
it seems Lesa’s (https://forum.openwrt.org/u/pesa1234) latest build has finally resolved the bridger issues i was experiencing with my APs not giving internet to my devices anymore.

**mt76: imported new mtk patch** 
do not report ACK when TXS is lost 
refine twt mcu update flow

Seems like this patch doesn't break bridger anymore. I am wondering if others can test. I have been up for 2 hours without issue. Usually only had 2-5 mins before bridger would break.

https://github.com/pesa1234/MT6000_cust_build/tree/main/2024.10.24_r28037_6.6.58_next-r4.3.6.mtk

**Comment by d687r02j8g** [comment]
> it seems Pesa’s (https://forum.openwrt.org/u/pesa1234) latest build has finally resolved the bridger issues i was experiencing with my APs not giving internet to my devices anymore.
> 
> **mt76: imported new mtk patch** do not report ACK when TXS is lost refine twt mcu update flow
> 
> Seems like this patch doesn't break bridger anymore. I am wondering if others can test. I have been up for 2 hours without issue. Usually only had 2-5 mins before bridger would break.
> 
> https://github.com/pesa1234/MT6000_cust_build/tree/main/2024.10.24_r28037_6.6.58_next-r4.3.6.mtk

Are these patches eventually available in the main openwrt code? I have the same issue with bridger/wed on 24.10.0

**Comment by gssjshark** [comment]
> > it seems Pesa’s (https://forum.openwrt.org/u/pesa1234) latest build has finally resolved the bridger issues i was experiencing with my APs not giving internet to my devices anymore.
> > **mt76: imported new mtk patch** do not report ACK when TXS is lost refine twt mcu update flow
> > Seems like this patch doesn't break bridger anymore. I am wondering if others can test. I have been up for 2 hours without issue. Usually only had 2-5 mins before bridger would break.
> > https://github.com/pesa1234/MT6000_cust_build/tree/main/2024.10.24_r28037_6.6.58_next-r4.3.6.mtk
> 
> Are these patches eventually available in the main openwrt code? I have the same issue with bridger/wed on 24.10.0

I don’t think so, you can try the custom build from Pesa or use his patches and build your own.

**Comment by imwhocodes** [comment]
I'm having the same problem: stalled connections on AP migration
This is happening on multiple dumb AP ([RT3200](https://openwrt.org/toh/linksys/e8450)) with WED enabled (+ bridger)

> **mt76: imported new mtk patch**
> do not report ACK when TXS is lost
> refine twt mcu update flow

@nbd168 Can this patch be added to your upstream https://github.com/openwrt/mt76 ?

Thanks,
Luca

**Comment by gssjshark** [comment]
It seems this patch fixes the issue:

https://github.com/pesa1234/openwrt/blob/next-r4.5.34.rss.mtk/target/linux/generic/hack-6.6/999-do-not-disable-mt7530-sa-learn.patch

**Comment by gssjshark** [comment]
It stopped working after a while for me as well

**Comment by Headcrabed** [comment]
Anyone could try these two patches?
https://github.com/ericwoud/linux/commit/a73a2105684f3f738a97997ccc8f48637b1acb6b
https://github.com/ericwoud/linux/commit/f28c2e09ad19331e2f9beec4fc1b01cc013647e1

**Comment by gssjshark** [comment]
> Please test the latest version, it should work now.

hi @nbd168,

so i thought it was working but noticed 
cat /sys/kernel/debug/ppe0/bind
is always empty so not entirely sure if offloading is actually happening

**Comment by d687r02j8g** [comment]
> Please test the latest version, it should work now.

I have tried the bridger package on 25.12.4 on a dumbap and the issue still remains when roaming between access points.

@gssjshark - is this still an issue for you?
I can see data in /bind but roaming still breaks access to the WiFi.

**Comment by gssjshark** [comment]
> > Please test the latest version, it should work now.
> 
> I have tried the bridger package on 25.12.4 on a dumbap and the issue still remains when roaming between access points.
> 
> @gssjshark - is this still an issue for you?
> I can see data in /bind but roaming still breaks access to the WiFi.

Yes still an issue.

**Comment by d687r02j8g** [comment]
FYI - this pull request (https://github.com/openwrt/openwrt/pull/24038) eliminates the need for bridger, offloading is working in AP mode. I am currently testing it.
